### PR TITLE
Single-path symex checker [blocks: #3976]

### PR DESCRIPTION
**Pull request by peterschrammel**
Implements the --paths variant of using goto-symex.

- [X] Each commit message has a non-empty body, explaining why the change was made.
- [X] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [X] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 0420c96f4c3c483e0f62e1bed5f79726f148fffc).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98863639
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

File: src/goto-checker/bmc_util.h [hunk 1]
```diff
@@ -69,6 +69,12 @@ void slice(
   const optionst &,
   ui_message_handlert &);
 
+void output_coverage_report(
```
Comment (tautschnig) [inline src/goto-checker/bmc_util.h:100]: Documentation would be appreciated. I think the `cov_out` parameter is the one needing it the most, notably the aspect that an empty string will mean that nothing happens. How about:
```
/// Output a coverage report as generated by \ref symex_coveraget if \p cov_out is non-empty.
/// \param cov_out: file to write the report to; no report is generated if this is empty
/// \param goto_model: goto model to build the coverage report for
/// \param symex: symbolic execution run to report coverage for
/// \param ui_message_handler: status/warning message handler
```

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: f94473badd9eb23d35ae76594ebb9ce6d7fcccbe).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99520817

File: src/goto-checker/bmc_util.cpp [hunk 2]
```diff
@@ -285,3 +285,54 @@ void output_coverage_report(
...
+static void postprocess_equation(
+  symex_bmct &symex,
+  symex_target_equationt &equation,
+  const optionst &options,
+  const namespacet &ns,
+  ui_message_handlert &ui_message_handler)
+{
+  // add a partial ordering, if required
+  // We won't be able to decide any properties by adding this,
+  // but we'd like to see the entire SSA.
```
Comment (tautschnig) [inline src/goto-checker/bmc_util.cpp]: Does this comment still make sense? I recall requesting it, but now that this is more generic I don't think it's appropriate anymore.

File: src/goto-checker/bmc_util.cpp [hunk 3]
```diff
@@ -285,3 +285,54 @@ void output_coverage_report(
...
+  const namespacet &ns,
+  ui_message_handlert &ui_message_handler)
+{
+  auto get_goto_function =
+    [&goto_model](
+      const irep_idt &id) -> const goto_functionst::goto_functiont & {
+    return goto_model.get_goto_function(id);
+  };
+
+  // perform symbolic execution
```
Comment (tautschnig) [inline src/goto-checker/bmc_util.cpp]: Comments don't always add value: I think it's pretty clear what the next line does, and this just makes it sound like something surprising would happen?

File: src/goto-checker/bmc_util.cpp [hunk 4]
```diff
@@ -285,3 +285,54 @@ void output_coverage_report(
...
+  symex_target_equationt &equation,
+  const optionst &options,
+  const namespacet &ns,
+  ui_message_handlert &ui_message_handler)
+{
+  auto get_goto_function =
+    [&goto_model](
+      const irep_idt &id) -> const goto_functionst::goto_functiont & {
+    return goto_model.get_goto_function(id);
+  };
```
Comment (tautschnig) [inline src/goto-checker/bmc_util.cpp:346]: Should we maybe tweak the `goto_symext` interface a bit? Looking at `get_function_from_goto_functions` (`symex_main.cpp`) we now define this helper function multiple times.

File: src/goto-checker/bmc_util.cpp [hunk 5]
```diff
@@ -336,3 +336,25 @@ void perform_symex(
 
   postprocess_equation(symex, equation, options, ns, ui_message_handler);
 }
+
+void perform_symex(
```
Comment (tautschnig) [inline src/goto-checker/bmc_util.cpp:353]: I'm confused. It seems two commits now introduce the same. They actually do, the lines above already contain the same code.

File: src/goto-checker/bmc_util.h [hunk 6]
```diff
@@ -79,6 +80,16 @@ void perform_symex(
   const namespacet &,
   ui_message_handlert &);
 
+/// Perform symbolic execution starting from \p resume.
+void perform_symex(
```
Comment (tautschnig) [inline src/goto-checker/bmc_util.h:84]: Declaration is also added twice.

File: src/goto-checker/bmc_util.h [hunk 7]
```diff
@@ -117,6 +117,16 @@ void update_status_of_not_checked_properties(
   propertiest &properties,
   std::unordered_set<irep_idt> &updated_properties);
 
+/// Sets the property status of UNKNOWN properties to PASS.
+/// \param [inout] properties: The status is updated in this data structure
```
Comment (tautschnig) [inline src/goto-checker/bmc_util.h]: The coding standard uses `in,out`

File: src/goto-checker/goto_symex_property_decider.cpp [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+  const namespacet &ns)
+  : options(options), ui_message_handler(ui_message_handler), equation(equation)
+{
+  solver_factoryt solvers(
+    options,
+    ns,
+    ui_message_handler,
+    ui_message_handler.get_ui() == ui_message_handlert::uit::XML_UI);
+  solver = solvers.get_solver();
+  solver->prop_conv().set_message_handler(ui_message_handler);
```
Comment (tautschnig) [inline src/goto-checker/goto_symex_property_decider.cpp]: That should really be the `solver_factoryt`'s job. Invoking `get_solver()` ought to make sure the message handler is set up, the factory does have all the necessary info.

File: src/goto-checker/goto_symex_property_decider.cpp [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+}
+
+exprt goto_symex_property_decidert::goalt::as_expr() const
+{
+  std::vector<exprt> tmp;
```
Comment (tautschnig) [inline src/goto-checker/goto_symex_property_decider.cpp]: `exprt::operandst conjuncts;`

File: src/goto-checker/Makefile [hunk 10]
```diff
@@ -7,6 +7,7 @@ SRC = bmc_util.cpp \
       goto_verifier.cpp \
       multi_path_symex_checker.cpp \
       multi_path_symex_only_checker.cpp \
+      single_path_symex_only_checker.cpp \
```
Comment (tautschnig) [inline src/goto-checker/Makefile]: Lexicographic sorting, please. Applies again in a later commit (introducing `single_path_symex_checker.cpp`).

File: src/cbmc/cbmc_parse_options.cpp [hunk 11]
```diff
@@ -559,13 +560,22 @@ int cbmc_parse_optionst::doit()
...
-      return CPROVER_EXIT_SUCCESS;
     }
+    else
+    {
+      verifier = util_make_unique<
+        all_properties_verifiert<multi_path_symex_only_checkert>>(
+        options, ui_message_handler, goto_model);
+    }
+
+    (void)(*verifier)();
```
Comment (tautschnig) [inline src/cbmc/cbmc_parse_options.cpp]: Seems like an unnecessary dance, can't we just run each of them in the branches above?

File: src/cbmc/cbmc_parse_options.cpp [hunk 12]
```diff
@@ -581,13 +582,22 @@ int cbmc_parse_optionst::doit()
...
+          options, ui_message_handler, goto_model);
+    }
+    else
+    {
+      verifier =
+        util_make_unique<stop_on_fail_verifiert<multi_path_symex_checkert>>(
+          options, ui_message_handler, goto_model);
     }
+
+    (void)(*verifier)();
```
Comment (tautschnig) [inline src/cbmc/cbmc_parse_options.cpp]: Again, seems unnecessary. Just call each of them in the branches above.

File: jbmc/src/jbmc/jbmc_parse_options.cpp [hunk 13]
```diff
@@ -559,46 +561,83 @@ int jbmc_parse_optionst::doit()
...
+      else
+      {
+        verifier = util_make_unique<
+          all_properties_verifiert<java_multi_path_symex_only_checkert>>(
           options, ui_message_handler, goto_model);
-        (void)verifier();
-        return CPROVER_EXIT_SUCCESS;
       }
+
+      (void)(*verifier)();
```
Comment (tautschnig) [inline jbmc/src/jbmc/jbmc_parse_options.cpp]: Just call each of them in the branches above.

File: jbmc/src/jbmc/jbmc_parse_options.cpp [hunk 14]
```diff
@@ -559,46 +561,83 @@ int jbmc_parse_optionst::doit()
...
-        return CPROVER_EXIT_SUCCESS;
       }
+      else
+      {
+        verifier = util_make_unique<
+          stop_on_fail_verifiert<java_multi_path_symex_checkert>>(
+          options, ui_message_handler, goto_model);
+      }
+
+      (void)(*verifier)();
```
Comment (tautschnig) [inline jbmc/src/jbmc/jbmc_parse_options.cpp]: Just call each of them.

**Review by tautschnig — APPROVED** [review]

File: src/goto-checker/bmc_util.cpp (hunk 5 above)
Comment (peterschrammel) [reply src/goto-checker/bmc_util.cpp:353]: Yes, that's exactly the problem.  The goto-symex interface is broken in this respect. We should have a function that returns a `path_storaget::patht` for the initial state at the entry point; and then we can `resume` from that... and everything becomes nice and uniform.

Comment (tautschnig) [reply src/goto-checker/bmc_util.cpp:353]: That should be fixable?

**Review by smowton — APPROVED** [review]

File: src/goto-checker/bmc_util.cpp (hunk 5 above)
Comment (peterschrammel) [reply src/goto-checker/bmc_util.cpp:353]: Yes, but I'll go ahead with the current version for now and clean up the goto-symex interface later this week. https://github.com/diffblue/cbmc/issues/4071

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 1f9f4756d8eb7f6fba819ba2d3913fca3517a61e).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99702092

**Pull request merged**